httputil/write_interceptor.go

```diff
@@ -0,0 +1,34 @@
+package httputil
+
+import "net/http"
+
+func NewHttpWriteInterceptor(w http.ResponseWriter) *HttpWriteInterceptor {
```

Comment:
There are several other interfaces that the actual default ResponseWriter implements and are needed to handle things like CloseNotify or Upgrade. CloseNotify is important for logs streaming, for example. Without a fuller implementation of those optional interfaces, you need to be very thoughtful about where you use this to ensure those endpoints do not need those interfaces.

Reply:
That's a good point you raise.

Whilst that may be the case, this code is already being used in production by all OpenFaaS users and exercised through the path: gateway => provider.

For that reason I wonder whether that code is required for our use-cases.

https://github.com/openfaas/faas/blob/52d0165c3f22b5d1c269dd6e43e6e6bda6559936/gateway/handlers/notifier_handler.go#L16

Reply:
It's current usage is not nearly as extensive as this would be 
![image](https://user-images.githubusercontent.com/891889/175994258-111f692c-db25-4982-90b3-1217f9831835.png)


As far as I can tell it is only used for `faasHandlers.Alert`, `QueuedProxy`, and `AsyncReport`. I think the nature of those handlers is pretty short typical requests, whic his why there are no issues.

Reply:
These are all also very short lived requests to the Kubernetes API to get JSON, and the function proxy is unaffected.

The logs endpoint may run for longer, however that is already being wrapped with an auth handler in a similar way in the gateway.

I'm not sure what you would like me to change, so can run the test image with the logs command and verify if there is an issue to report?

Reply:
It seems like `CloseNotifier` may be deprecated? https://pkg.go.dev/net/http#CloseNotifier

Can you show a use-case where this will fail in the way you're describing?

Reply:
> On a related note, it seems like `CloseNotifier` may be deprecated? https://pkg.go.dev/net/http#CloseNotifier
> 
> I got the error you mentioned: `2022/06/30 09:14:15 LogHandler: response is not a CloseNotifier, required for streaming response`
> 
> Is there no way to have these additional behaviors simply proxied to the underlying object, or is a third party dependency on CHI etc required?

I don't understand how this is a response to my previous statement.

1) I link directly to a usage in the current code base where this is going to break. I don't understand the request for yet another example.
2) The link to an example implementation is not a statement for "we need to use this". I thought my previous comment was pretty clear, the implementation (either implemented internally or from a new dependency) needs to have this Flusher implementation, that is all. The link was simply to give an example of an existing implementation that can be used as a model. If you look at that implementation you will see that the wrapper just delegates to the original flusher, see https://github.com/go-chi/chi/blob/9db25e5b381fbd6350583b857e74ecaffe355819/middleware/wrap_writer.go#L125-L129. I don't care how it is implemented, but it _needs_ to be implemented if you want streaming logs to continue to work.
3) Regarding the CloseNotifier, yes that is true. It simply needs to be updated in our implementation. If you don't want that in this ResponseWriter wrapper, then you will need to update the log handler. If you don't want to do the update in the handler, then the wrapper needs to implement the CloseNotifier as well.

Reply:
It seems like we are talking over each other, which is not helping.

I was asking you what problem would occur, in order to learn. And because I hadn't used the log endpoint in testing, I didn't know what error we would run into. So when I tried the log handler, I found the error for myself, and shared it.

I've resolved this now with an updated version.